python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -0,0 +1,260 @@
+from tensorflow import lookup
+
+import copy
+
+from ml4ir.base.io import file_io
+
+
+def categorical_embedding_with_hash_buckets(feature_tensor, feature_info):
+    """
+    Convert a string feature tensor into a categorical embedding.
+    Works by first converting the string into num_hash_buckets buckets
+    each of size hash_bucket_size, then converting each hash bucket into
+    a categorical embdding of dimension embedding_size. Finally, these embeddings
+    are combined either through mean, sum or concat operations to generate the final
+    embedding.
+
+    Args:
+        feature_tensor: String feature tensor
+        feature_info: Dictionary representing the feature_info for the specific feature from the FeatureConfig
+
+    Returns:
+        categorical embedding for the input feature_tensor
+
+    Args under feature_layer_info:
+        num_hash_buckets: int; number of different hash buckets to convert the input string into
+        hash_bucket_size: int; the size of each hash bucket
+        embedding_size: int; dimension size of the categorical embedding
+        merge_mode: str; can be one of "mean", "sum", "concat" representing the mode of combining embeddings from each categorical embedding
+    """
+
+    # Numeric input features
+    if feature_info["dtype"] in (tf.float32, tf.int64):
+        raise NotImplementedError
+
+    # String input features
+    elif feature_info["dtype"] in (tf.string,):
+        feature_layer_info = feature_info.get("feature_layer_info")
+        embeddings_list = list()
+        for i in range(feature_layer_info["args"]["num_hash_buckets"]):
+            augmented_string = layers.Lambda(lambda x: tf.add(x, str(i)))(feature_tensor)
+
+            hash_bucket = tf.strings.to_hash_bucket_fast(
+                augmented_string, num_buckets=feature_layer_info["args"]["hash_bucket_size"]
+            )
+            embeddings_list.append(
+                layers.Embedding(
+                    input_dim=feature_layer_info["args"]["hash_bucket_size"],
+                    output_dim=feature_layer_info["args"]["embedding_size"],
+                    name="categorical_embedding_{}_{}".format(feature_info.get("name"), i),
+                )(hash_bucket)
+            )
+
+        embedding = None
+        if feature_layer_info["args"]["merge_mode"] == "mean":
+            embedding = tf.reduce_mean(
+                embeddings_list,
+                axis=0,
+                name="categorical_embedding_{}".format(feature_info.get("name")),
+            )
+        elif feature_layer_info["args"]["merge_mode"] == "sum":
+            embedding = tf.reduce_sum(
+                embeddings_list,
+                axis=0,
+                name="categorical_embedding_{}".format(feature_info.get("name")),
+            )
+        elif feature_layer_info["args"]["merge_mode"] == "concat":
+            embedding = tf.concat(
+                embeddings_list,
+                axis=-1,
```

Comment:
out of curiosity, why do we have a "axis=-1" for this one?

Reply:
Because it is concatenating on the last axis which has the embeddings, whereas the other operations are reduce operations on the first axis. It is a strange tensorflow usage.

```
In [20]: tf.concat([np.random.randn(128, 16)] * 4, axis=1).shape
Out[20]: TensorShape([128, 64])

In [21]: tf.reduce_sum([np.random.randn(128, 16)] * 4, axis=0).shape
Out[21]: TensorShape([128, 16])
```